File: sdks/go/computer_test.go
package fuse

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestComputer(t *testing.T) {
	var gotPath, gotMethod string
	var gotBody []byte
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.Path
		gotMethod = r.Method
		gotBody, _ = io.ReadAll(r.Body)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"screenshot":"cGl4ZWxz"}`))
	}))
	defer srv.Close()

	c, err := New(srv.URL, "tok")
	if err != nil {
		t.Fatal(err)
	}
	res, err := c.Environments.Computer(context.Background(), "vm-1", ComputerActionRequest{
		Action:     "left_click",
		Coordinate: []int{100, 200},
	})
	if err != nil {
		t.Fatalf("Computer: %v", err)
	}
	if gotMethod != http.MethodPost || gotPath != "/v1/environments/vm-1/computer" {
		t.Fatalf("request = %s %s, want POST /v1/environments/vm-1/computer", gotMethod, gotPath)
	}
	var sent ComputerActionRequest
	if err := json.Unmarshal(gotBody, &sent); err != nil {
		t.Fatalf("sent body is not json: %v", err)
	}
	if sent.Action != "left_click" || len(sent.Coordinate) != 2 {
		t.Fatalf("sent = %+v", sent)
	}
	if res.Screenshot != "cGl4ZWxz" {
		t.Fatalf("screenshot = %q", res.Screenshot)
	}
}

func TestComputerDisplay(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/environments/vm-1/computer" || r.Method != http.MethodGet {
			t.Errorf("request = %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"display":":1","up":true,"width":1280,"height":800}`))
	}))
	defer srv.Close()

	c, err := New(srv.URL, "tok")
	if err != nil {
		t.Fatal(err)
	}
	res, err := c.Environments.ComputerDisplay(context.Background(), "vm-1")
	if err != nil {
		t.Fatalf("ComputerDisplay: %v", err)
	}
	if !res.Up || res.Width != 1280 || res.Height != 800 {
		t.Fatalf("display = %+v", res)
	}
}

func TestComputerValidation(t *testing.T) {
	c, err := New("http://127.0.0.1:0", "tok")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := c.Environments.Computer(context.Background(), "", ComputerActionRequest{Action: "screenshot"}); err == nil {
		t.Fatal("empty vm id must be rejected before any request is made")
	}
	if _, err := c.Environments.Computer(context.Background(), "vm-1", ComputerActionRequest{}); err == nil {
		t.Fatal("empty action must be rejected before any request is made")
	}
	if _, err := c.Environments.ComputerDisplay(context.Background(), ""); err == nil {
		t.Fatal("empty vm id must be rejected before any request is made")
	}
}

File: sdks/typescript/test/computer.test.ts
import { afterEach, describe, expect, it } from "vitest";
import { serve, readBody, pathOf, type TestServer } from "./server.js";

let current: TestServer | undefined;
afterEach(async () => {
  await current?.close();
  current = undefined;
});

describe("environments.computer", () => {
  it("posts the action to the computer sub-path and returns the result", async () => {
    let seenPath = "";
    let seenMethod = "";
    let seenBody: unknown;
    current = await serve(async (req, res) => {
      seenPath = pathOf(req);
      seenMethod = req.method ?? "";
      seenBody = JSON.parse(await readBody(req));
      res.setHeader("content-type", "application/json");
      res.end(JSON.stringify({ screenshot: "cGl4ZWxz" }));
    });

    const res = await current.client.environments.computer("vm-1", {
      action: "left_click",
      coordinate: [100, 200],
    });

    expect(seenMethod).toBe("POST");
    expect(seenPath).toBe("/v1/environments/vm-1/computer");
    expect(seenBody).toEqual({ action: "left_click", coordinate: [100, 200] });
    expect(res.screenshot).toBe("cGl4ZWxz");
  });

  it("reads the display report from the same sub-path", async () => {
    current = await serve(async (req, res) => {
      expect(req.method).toBe("GET");
      expect(pathOf(req)).toBe("/v1/environments/vm-1/computer");
      res.setHeader("content-type", "application/json");
      res.end(JSON.stringify({ display: ":1", up: true, width: 1280, height: 800 }));
    });

    const res = await current.client.environments.computerDisplay("vm-1");
    expect(res.up).toBe(true);
    expect(res.width).toBe(1280);
  });

  it("rejects an empty vm id or action before any request is made", async () => {
    current = await serve(async () => {
      throw new Error("no request should be made");
    });
    await expect(
      current.client.environments.computer("", { action: "screenshot" }),
    ).rejects.toThrow("vm id is required");
    await expect(
      current.client.environments.computer("vm-1", { action: "" }),
    ).rejects.toThrow("action is required");
  });
});
